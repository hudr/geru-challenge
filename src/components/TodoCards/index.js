import React, { Component, Fragment } from 'react'
import PropTypes from 'prop-types'

import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'

import format from 'date-fns/format'

import * as TodoActions from '../../redux/actions/todo'

import {
  StyledContainer,
  CardContainer,
  ContainerTitle,
  CardTitle,
  ContainerActionButtons,
  ActionButton,
  CardDescription,
  InformationContainer,
  CardStatus,
  TagContainer,
  CardTag,
  StyledFilterContainer,
  StyledForm
} from './styled'

class TodoCards extends Component {
  constructor(props) {
    super(props)
    this.state = {
      done: 'All',
      myTodos: []
    }
  }

  handleStatusChange(e) {
    this.setState({
      done: e.target.value
    })
  }

  render() {
    const { todos, openModal, removeTodo } = this.props
    const { done } = this.state

    const todoAux =
      done === 'All' ? [...todos] : todos.filter(todo => todo.done === done)

    return (
      <Fragment>
        <StyledFilterContainer>
          <StyledForm>
            <select onChange={this.handleStatusChange.bind(this)} value={done}>
              <option value='All'>All</option>
              <option value='false'>Pending</option>
              <option value='true'>Done</option>
            </select>
          </StyledForm>
        </StyledFilterContainer>
        <StyledContainer>
          {todoAux.map(todo => (
            <CardContainer key={todo.id}>
              <ContainerTitle>
                <CardTitle>{todo.title}</CardTitle>
                <ContainerActionButtons>
                  <ActionButton onClick={() => openModal('put', todo)}>
                    Edit
                  </ActionButton>
                  <ActionButton
                    bgColor='#cc0000'
                    onClick={() => removeTodo(todo.id)}
                  >
                    Remove
                  </ActionButton>
                </ContainerActionButtons>
              </ContainerTitle>
              <CardDescription>{todo.description}</CardDescription>
              <InformationContainer>
                <div>
                  <div>
                    <strong>Status: </strong>
                    {todo.done === 'true' ? (
                      <CardStatus fontColor={'green'}>Done!</CardStatus>
                    ) : (
                      <CardStatus>Pending.</CardStatus>
                    )}
                  </div>
                  <div>
                    <strong>Created:</strong>{' '}
                    {format(new Date(todo.createDate), 'dd/MM/yy HH:mm')}
                  </div>
                </div>
                <div>
                  <div>
                    <strong>Est.:</strong> {todo.estTime} day(s).
                  </div>
                  <div>
                    <strong>Due:</strong>{' '}
                    {format(new Date(todo.dueDate), 'dd/MM/yy HH:mm')}
                  </div>
                </div>
              </InformationContainer>
              <TagContainer>
                {todo.tags.map((tag, i) => (
                  <CardTag key={i}>#{tag.toLowerCase()}</CardTag>
                ))}
              </TagContainer>
            </CardContainer>
          ))}
        </StyledContainer>
      </Fragment>
    )
  }
}

const mapStateToProps = state => ({
  todos: state.todo.todos,
  modalType: state.todo.modalType
})

const mapDispatchToProps = dispatch => bindActionCreators(TodoActions, dispatch)

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(TodoCards)

TodoCards.propTypes = {
  todos: PropTypes.array.isRequired,
  modalType: PropTypes.string
}

export { TodoCards }
