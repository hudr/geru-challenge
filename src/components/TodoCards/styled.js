import styled from 'styled-components'

export const StyledFilterContainer = styled.div`
  margin-top: 10vh;
  width: 100%;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
`

export const StyledForm = styled.form`
  width: 95%;

  border: 1px solid #ccc;
  border-radius: 3px 3px 0px 0px;
  overflow: hidden;
  background: #fafafa;

  select {
    padding: 10px 13px;
    width: 100%;
    border: none;
    box-shadow: none;
    background: transparent;
    background-image: none;
    -webkit-appearance: none;

    text-align-last: center;
  }

  select:focus {
    outline: none;
  }
`

export const StyledContainer = styled.div`
  margin-top: 2vh;
  display: flex;
  justify-content: center;
  align-items: center;
  flex-direction: column;
`

export const CardContainer = styled.div`
  width: 95%;
  margin-bottom: 2vh;
  border-radius: 4px;
  background: #fff;
  box-shadow: 0 6px 10px rgba(0, 0, 0, 0.08), 0 0 6px rgba(0, 0, 0, 0.05);
  transition: 0.3s transform cubic-bezier(0.155, 1.105, 0.295, 1.12),
    0.3s box-shadow,
    0.3s -webkit-transform cubic-bezier(0.155, 1.105, 0.295, 1.12);
  cursor: pointer;

  :hover {
    transform: scale(1.01);
    box-shadow: 0 10px 20px rgba(0, 0, 0, 0.12), 0 4px 8px rgba(0, 0, 0, 0.06);
  }
`
export const ContainerTitle = styled.div`
  border-top-left-radius: 4px;
  border-top-right-radius: 4px;
  background-color: #03adfc;
  padding: 5px 0px 5px 20px;

  display: flex;
  justify-content: space-between;
`

export const CardTitle = styled.h3`
  color: white;
  font-size: 20px;
  font-weight: 600;
  padding-top: 4px;
`

export const ContainerActionButtons = styled.div`
  padding: 5px 20px 5px 0px;
`

export const ActionButton = styled.button`
  border: 0;
  border-radius: 5px;
  width: 45px;
  height: 30px;
  background: ${props => (props.bgColor ? props.bgColor : '#4776e6')};
  cursor: pointer;
  font-size: 0.69rem;
  color: #fff;
  margin-left: 10px;
`

export const CardDescription = styled.p`
  padding: 14px 80px 18px 30px;
  color: grey;
  font-size: 0.94rem;
`

export const InformationContainer = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: space-between;

  align-items: center;
  font-size: 13px;

  padding: 10px 30px 10px 30px;
  color: #3d3d3d;
  background-color: rgba(3, 173, 252, 0.2);
`

export const CardStatus = styled.span`
  font-weight: bold;
  color: ${props => (props.fontColor ? props.fontColor : '#cc0000')};
`

export const TagContainer = styled.div`
  margin-top: 10px;
  display: flex;
  justify-content: center;
  margin-bottom: 10px;
`

export const CardTag = styled.span`
  border-radius: 30px;
  margin-right: 10px;
  color: ${props => (props.fontColor ? props.fontColor : '#03adfc')};
  font-size: 10px;
  font-style: italic;
`

export const CardNoTag = styled.span`
  border-radius: 30px;
  margin-right: 10px;
  color: #cc0000;
  font-size: 10px;
  font-style: italic;
`
