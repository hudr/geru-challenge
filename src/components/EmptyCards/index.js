import React, { Component, Fragment } from 'react'

import { StyledEmptyContainer, NotesImg, Subtitle } from './styled'

import emptyImage from '../../assets/images/nodata.svg'

class EmptyCards extends Component {
  render() {
    return (
      <Fragment>
        <StyledEmptyContainer>
          <NotesImg src={emptyImage} alt='emptyimage' />
          <Subtitle>Your todo list is empty :)</Subtitle>
        </StyledEmptyContainer>
      </Fragment>
    )
  }
}

export default EmptyCards
