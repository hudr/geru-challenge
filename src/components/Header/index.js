import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'

import * as TodoActions from '../../redux/actions/todo'

import { StyledContainer, Title, CreateButton } from './styled'

class Header extends Component {
  render() {
    const { openModal } = this.props
    return (
      <StyledContainer>
        <Title>Geru Challenge</Title>
        <CreateButton onClick={() => openModal('post')}>Create</CreateButton>
      </StyledContainer>
    )
  }
}

const mapDispatchToProps = dispatch => bindActionCreators(TodoActions, dispatch)

export default connect(
  null,
  mapDispatchToProps
)(Header)

Header.propTypes = {
  openModal: PropTypes.func
}

export { Header }
