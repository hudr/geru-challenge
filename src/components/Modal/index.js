import React, { Component, Fragment } from 'react'
import PropTypes from 'prop-types'

import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'

import * as TodoActions from '../../redux/actions/todo'

import Button from '@material-ui/core/Button'
import Dialog from '@material-ui/core/Dialog'
import DialogActions from '@material-ui/core/DialogActions'
import DialogContent from '@material-ui/core/DialogContent'
import DialogContentText from '@material-ui/core/DialogContentText'
import DialogTitle from '@material-ui/core/DialogTitle'
import { withStyles } from '@material-ui/core/styles'
import withMobileDialog from '@material-ui/core/withMobileDialog'
import TextField from '@material-ui/core/TextField'
import Radio from '@material-ui/core/Radio'
import RadioGroup from '@material-ui/core/RadioGroup'
import FormControlLabel from '@material-ui/core/FormControlLabel'

import { MuiPickersUtilsProvider, DateTimePicker } from '@material-ui/pickers'
import DateFnsUtils from '@date-io/date-fns'

import TagsInput from 'react-tagsinput'
import 'react-tagsinput/react-tagsinput.css'

import { TagsContainer } from './styled'

const useStyles = theme => ({
  textField: {
    marginLeft: theme.spacing(0),
    marginRight: theme.spacing(0)
  },
  dateField: {
    marginTop: theme.spacing(4)
  },
  radioButton: {
    marginTop: theme.spacing(2)
  }
})

class Modal extends Component {
  constructor(props) {
    super(props)
    this.state = {
      id: props.currentTodo.id || '',
      title: props.currentTodo.title || '',
      description: props.currentTodo.description || '',
      dueDate: props.currentTodo.dueDate || new Date(),
      createDate: props.currentTodo.createDate || new Date(),
      estTime: props.currentTodo.estTime || '',
      done: props.currentTodo.done || 'false',
      tags: props.currentTodo.tags || []
    }

    this.handleTagsChange = this.handleTagsChange.bind(this)
  }

  handleDateChange = date => {
    this.setState({
      dueDate: date
    })
  }

  handleTagsChange(tags) {
    this.setState({ tags })
  }

  handleStatusChange(event) {
    this.setState({ done: event.target.value })
  }

  handleInputOnChange = e => {
    this.setState({ [e.target.name]: e.target.value })
  }

  handleSubmit = async event => {
    event.preventDefault()

    const { modalType, addTodo, updateTodo } = this.props

    const {
      id,
      title,
      description,
      createDate,
      dueDate,
      estTime,
      done,
      tags
    } = this.state

    const currentTodo = {
      id,
      title,
      description,
      createDate,
      dueDate,
      estTime,
      done,
      tags
    }

    if (modalType === 'put') {
      await updateTodo(currentTodo)
    } else {
      await addTodo(currentTodo)
    }

    return this.setState({
      title: '',
      description: '',
      estTime: '',
      dueDate,
      tags: []
    })
  }

  render() {
    const {
      fullScreen,
      classes,
      handleClose,
      isModalOpen,
      closeModal,
      modalType
    } = this.props

    const { title, description, estTime, dueDate, done } = this.state
    return (
      <Fragment>
        <Dialog
          fullScreen={fullScreen}
          open={isModalOpen}
          onClose={handleClose}
          aria-labelledby='responsive-dialog-title'
        >
          <DialogTitle id='responsive-dialog-title'>
            {'Fill your todo info :)'}
          </DialogTitle>
          <DialogContent>
            <DialogContentText>
              Don't forget to add tags, they're important to remember what you
              need to do. Tag names can't be the same.
            </DialogContentText>
            <form
              onSubmit={this.handleSubmit.bind(this)}
              onChange={this.handleInputOnChange}
              id='todoinfo'
              autoComplete='off'
            >
              <TextField
                label='Title'
                required
                name='title'
                placeholder='e.g.: Homework'
                value={title}
                multiline
                className={classes.textField}
                margin='normal'
                fullWidth
              />

              <TextField
                label='Description'
                placeholder='e.g.: Finish biology research'
                name='description'
                value={description}
                className={classes.textField}
                margin='normal'
                fullWidth
                required
              />

              <TextField
                label='Estimated Time in days'
                type='number'
                name='estTime'
                placeholder='e.g.: 3'
                value={estTime}
                className={classes.textField}
                InputProps={{ inputProps: { min: 1 } }}
                margin='normal'
                fullWidth
                required
              />

              {modalType === 'put' && (
                <Fragment>
                  <RadioGroup
                    onChange={e => this.handleStatusChange(e)}
                    aria-label='status'
                    name='status'
                    value={done}
                    className={classes.radioButton}
                  >
                    <FormControlLabel
                      control={<Radio color='primary' />}
                      label='Done'
                      name='true'
                      value='true'
                    />
                    <FormControlLabel
                      control={<Radio color='secondary' />}
                      label='Pending'
                      name='false'
                      value='false'
                    />
                  </RadioGroup>
                </Fragment>
              )}

              <MuiPickersUtilsProvider utils={DateFnsUtils}>
                <DateTimePicker
                  label='Due date:'
                  ampm={false}
                  value={dueDate}
                  onChange={this.handleDateChange}
                  className={classes.dateField}
                  animateYearScrolling={false}
                  required
                  disablePast
                  format={'dd/MM/yyyy HH:mm'}
                  showTodayButton
                  fullWidth
                />
              </MuiPickersUtilsProvider>

              <TagsContainer>
                <TagsInput
                  value={this.state.tags}
                  onlyUnique
                  maxTags={6}
                  inputProps={{
                    placeholder: 'Add a tag'
                  }}
                  onChange={this.handleTagsChange}
                />
              </TagsContainer>
            </form>
          </DialogContent>
          <DialogActions className={classes.actions}>
            <Button type='submit' form='todoinfo' color='primary'>
              {modalType === 'put' ? 'Update' : 'Insert'}
            </Button>
            <Button onClick={() => closeModal()} color='secondary'>
              Close
            </Button>
          </DialogActions>
        </Dialog>
      </Fragment>
    )
  }
}

const mapStateToProps = state => ({
  todos: state.todo.todos,
  isModalOpen: state.todo.isModalOpen,
  modalType: state.todo.modalType,
  currentTodo: state.todo.currentTodo
})

const mapDispatchToProps = dispatch => bindActionCreators(TodoActions, dispatch)

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(withMobileDialog()(withStyles(useStyles)(Modal)))

Modal.propTypes = {
  todos: PropTypes.array.isRequired,
  isModalOpen: PropTypes.bool.isRequired,
  modalType: PropTypes.string.isRequired,
  currentTodo: PropTypes.object.isRequired
}

export { Modal }
