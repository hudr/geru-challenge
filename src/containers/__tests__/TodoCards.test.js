import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import { TodoCards } from '../../components/TodoCards'

describe('<TodoCards />', () => {
  it('renders without crashing given props', () => {
    const props = {
      todos: [],
      removeTodo: jest.fn(),
      getTodos: jest.fn(),
      openModal: jest.fn(),
      isModalOpen: false
    }
    const wrapper = shallow(<TodoCards {...props} />)
    expect(toJson(wrapper)).toMatchSnapshot()
  })
})
