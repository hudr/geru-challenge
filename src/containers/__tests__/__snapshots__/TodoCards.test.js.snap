// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TodoCards /> renders without crashing given props 1`] = `
<Fragment>
  <styled.div>
    <styled.form>
      <select
        onChange={[Function]}
        value="All"
      >
        <option
          value="All"
        >
          All
        </option>
        <option
          value="false"
        >
          Pending
        </option>
        <option
          value="true"
        >
          Done
        </option>
      </select>
    </styled.form>
  </styled.div>
  <styled.div />
</Fragment>
`;
